server/lib/chat/chat-models.ts

```diff
@@ -381,18 +381,131 @@ export async function updateUserPermissions(
   }
 }
 
+export async function countBannedIdentifiersForChannel(
+  {
+    channelId,
+    targetId,
+    filterBrowserprint = true,
+  }: {
+    channelId: SbChannelId
+    targetId: SbUserId
+    filterBrowserprint?: boolean
+  },
+  withClient?: DbClient,
+): Promise<number> {
+  const { client, done } = await db(withClient)
+
+  try {
+    const query = sql`
+      SELECT COUNT(DISTINCT identifier_type) as "matches"
+      FROM channel_identifier_bans cib
+      WHERE cib.channel_id = ${channelId}
+      AND (cib.identifier_type, cib.identifier_hash) IN (
+        SELECT identifier_type, identifier_hash
+        FROM user_identifiers ui
+        WHERE ui.user_id = ${targetId}
+      )
+    `
+
+    if (filterBrowserprint) {
+      query.append(sql`
+        AND cib.identifier_type != 0
+      `)
+    }
+
+    const result = await client.query<{ matches: string }>(query)
+    return result.rows.length > 0 ? Number(result.rows[0].matches) : 0
+  } finally {
+    done()
+  }
+}
+
 export async function banUserFromChannel(
-  channelId: SbChannelId,
-  moderatorId: SbUserId,
-  targetId: SbUserId,
-  reason?: string,
+  {
+    channelId,
+    moderatorId,
+    targetId,
+    reason,
+    automated = false,
+    connectedUsers = [],
+  }: {
+    channelId: SbChannelId
+    moderatorId?: SbUserId
+    targetId: SbUserId
+    reason?: string
+    automated?: boolean
+    connectedUsers?: SbUserId[]
+  },
+  withClient?: DbClient,
 ): Promise<void> {
-  const { client, done } = await db()
+  const { client, done } = await db(withClient)
   try {
-    await client.query(sql`
-      INSERT INTO channel_bans (user_id, channel_id, ban_time, banned_by, reason)
-      VALUES (${targetId}, ${channelId}, ${new Date()}, ${moderatorId}, ${reason});
+    if (automated) {
+      await client.query(sql`
+        WITH rc AS (
+          SELECT reason, COUNT(*) AS reason_count
+          FROM channel_bans
+          WHERE channel_id = ${channelId} AND user_id = ANY(${connectedUsers})
+          GROUP BY reason
+          ORDER BY reason_count DESC
```

Comment:
probably want a secondary order here too so that ties are at least deterministic (ban date most likely)

Reply:
Hmm, adding ban date to the `ORDER BY` clause requires me to add it to the `GROUP BY` clause as well, I think, and I'm not sure if that's desirable?

Reply:
That doesn't seem to be the case given that `reason_count` is not in the `GROUP BY`, if you're getting an error here it's likely that it needs to appear in the `SELECT`. Since the goal is to pick the value with the highest `reason_count` and earliest date, you could likely add `MIN(date)` to the select and use that here